### PR TITLE
genchallange check has been added for Authorization

### PR DESCRIPTION
**Pull request by ikaratass**
fix [AB#2850](https://dev.azure.com/switch-ev/09ab882d-ca4b-4b5b-b5b8-80603c12e5b2/_workitems/edit/2850)

According to spec 15118-2  [V2G2-475], the GenChallenge in AuthorizationReq should be the same in PaymentDetailsRes. Otherwise, we need to return FAILED_ChallengeInvalid.

[V2G2-475] The message 'AuthorizationRes' shall contain the ResponseCode
'FAILED_ChallengeInvalid' if the challenge response contained in the
 AuthorizationReq message in attribute GenChallenge is not valid versus
the provided GenChallenge in PaymentDetailsRes.

### CODE REVIEW

File: iso15118/secc/states/iso15118_2_states.py [hunk 1]
```diff
@@ -1013,6 +1013,19 @@ async def process_message(
...
         authorization_req: AuthorizationReq = msg.body.authorization_req
 
+        # [V2G2-475] The message 'AuthorizationRes' shall contain the ResponseCode
+        # 'FAILED_ChallengeInvalid' if the challenge response contained in the
+        # AuthorizationReq message in attribute GenChallenge is not valid versus
+        # the provided GenChallenge in PaymentDetailsRes.
+        if authorization_req.gen_challenge:
+            if authorization_req.gen_challenge != self.comm_session.gen_challenge:
+                self.stop_state_machine(
+                    "GenChallenge is not the same in AuthorizationRes",
```
Comment (shalinnijel2) [inline iso15118/secc/states/iso15118_2_states.py:1023]: I believe you meant "GenChallenge not the same as what was sent in PaymentDetailsRes", right - not AuthorizationRes.

**Review by shalinnijel2 — CHANGES_REQUESTED** [review]

File: iso15118/secc/states/iso15118_2_states.py (hunk 1 above)
Comment (ikaratass) [reply iso15118/secc/states/iso15118_2_states.py]: it is changed as PaymentDetailsRes. Thanks...

File: tests/secc/states/test_iso15118_2_states.py [hunk 2]
```diff
@@ -193,6 +194,23 @@ async def test_authorization_next_state_on_authorization_request(
             == expected_evse_processing
         )
 
+    async def test_authorization_req_gen_challange(self):
+        self.comm_session.writer = Mock()
```
Comment (tropxy) [inline tests/secc/states/test_iso15118_2_states.py:198]: the test name should include the suffix invalid. Also I think you should create a test where the outcome is valid, if there's none yet

**Review by tropxy — CHANGES_REQUESTED** [review]

File: tests/secc/states/test_iso15118_2_states.py (hunk 2 above)
Comment (ikaratass) [reply tests/secc/states/test_iso15118_2_states.py]: Done

**Review by shalinnijel2 — DISMISSED** [review]

File: iso15118/secc/states/iso15118_2_states.py [hunk 3]
```diff
@@ -1013,6 +1013,19 @@ async def process_message(
...
         authorization_req: AuthorizationReq = msg.body.authorization_req
 
+        # [V2G2-475] The message 'AuthorizationRes' shall contain the ResponseCode
+        # 'FAILED_ChallengeInvalid' if the challenge response contained in the
+        # AuthorizationReq message in attribute GenChallenge is not valid versus
+        # the provided GenChallenge in PaymentDetailsRes.
+        if authorization_req.gen_challenge:
+            if authorization_req.gen_challenge != self.comm_session.gen_challenge:
+                self.stop_state_machine(
+                    "GenChallenge is not the same in PaymentDetailsRes",
```
[the remark this answers is not in the document]
Comment (ikaratass) [reply iso15118/secc/states/iso15118_2_states.py]: [V2G2-475] is added to in message.

File: iso15118/secc/states/iso15118_2_states.py [hunk 4]
```diff
@@ -1021,6 +1021,19 @@ async def process_message(
 
         authorization_req: AuthorizationReq = msg.body.authorization_req
 
+        # [V2G2-475] The message 'AuthorizationRes' shall contain the ResponseCode
+        # 'FAILED_ChallengeInvalid' if the challenge response contained in the
+        # AuthorizationReq message in attribute GenChallenge is not valid versus
+        # the provided GenChallenge in PaymentDetailsRes.
+        if authorization_req.gen_challenge:
```
Comment (tropxy) [inline iso15118/secc/states/iso15118_2_states.py:1028]: if the field gen_challenge is addressable but can be None, then do we need this if clause?
cant we just have the one after?

**Review by tropxy — CHANGES_REQUESTED** [review]

File: iso15118/secc/states/iso15118_2_states.py (hunk 4 above)
Comment (ikaratass) [reply iso15118/secc/states/iso15118_2_states.py:1028]: i see, gen_challenge is optional, if we remove this statement, we can face an issue when ev doesnt send gen_challenge value. if ev doesnt send it, the value will be None, and automatically we return as FAILED_CHALLENGE_INVALID, that is not a valid case. Am i wrong?

Comment (tropxy) [reply iso15118/secc/states/iso15118_2_states.py:1028]: you are not checking if the gen_challenge attribute exists in the authorization_req data structure. Instead, you are already assuming that the attribute exists but that its value can be None or not: this may be ok, depends how the pydantic model constructs the data. Is it the case that if the gen_challenge is None we will still be able to access the `gen_challenge`attribute? if so, then we dont need that first if clause, because in the one after we do the check as well

Comment (shalinnijel2) [reply iso15118/secc/states/iso15118_2_states.py:1028]: There is one use case being missed because of the if condition above - the case where GenChallenge is not send at all in AuthorizationReq in PnC.

I believe you added the check as the field is optional in EIM. It would continue to work in EIM if the check is removed anyway because of the reason @Andre mentioned above - but might throw an error in case a GenChallenge is accidentally sent in EIM AuthorizationReq. It may be better to ignore the field check in EIM completely?

```
if self.comm_session.selected_auth_option is AuthEnum.PNC_V2:
```

Comment (ikaratass) [reply iso15118/secc/states/iso15118_2_states.py:1028]: @shalinnijel2 should not we terminate the session with FAILED_CHALLENGE_INVALID in this case as well. I think, we might not need this check...

Comment (shalinnijel2) [reply iso15118/secc/states/iso15118_2_states.py]: Yes, true  - letting that pass through would also be a violation of the spec - in which case simply removing the if condition as discussed above would suffice.

File: tests/secc/states/test_messages.py [hunk 5]
```diff
@@ -118,11 +118,15 @@ def get_dummy_v2g_message_session_stop_req():
...
+    id: Optional[str] = None, gen_challenge: Optional[bytes] = None
+):
     # The AuthorizationReq is empty, unless it is following a PaymentDetailsRes
     # message, in which case it must send back the generated challenge.
-    authorization_req = AuthorizationReq()
-
+    if gen_challenge:
+        authorization_req = AuthorizationReq(id=id, gen_challenge=gen_challenge)
+    else:
+        authorization_req = AuthorizationReq()
```
Comment (tropxy) [inline tests/secc/states/test_messages.py:129]: if you are not generating here the challenge and just injecting, there is no point on having the if clause.
I would just pass along the injection, but is a minor thing

**Review by tropxy — APPROVED** [review]
